### DISCUSSION TITLE
Validation plugin

### DISCUSSION THREAD

**Discussion by mikekidner** [gaphor/gaphor#3259] [Ideas]
Has anyone considered how the validation of a model might be achieved? 
How to define rules? -> can a model be used to do this?
How users can modify/exclude etc the rules?
How to present results ?

The SAIC list of rules gives a starting point, (but are too perhaps specific to other tools) but would need to be 'translated' to refer to gaphor element attributes etc.  
[SAIC DE Validation Rules v26.xlsx](https://github.com/gaphor/gaphor/files/15032099/SAIC.DE.Validation.Rules.v26.xlsx)

**Comment by amolenaar** [comment]
Hi @mikekidner,

Validation can be done either inside the Gaphor ui, or as an external command. I think it depends on how you want validation to happen: instantly or on demand (e.g. a keyboard shortcut). Either way, I guess you'll have to implement the rules in Python (or build an interpreter in python).

I think it's a good idea to be able to run the validation from the command line in any case, so you can create a CI job and validate the model.

To include a validation service in the (ui) application you define one as an entry point. You can do validation either instantly (like [sanitizer service](https://github.com/gaphor/gaphor/blob/main/gaphor/UML/sanitizerservice.py) does, or as a command (like the [hello world plugin](https://github.com/gaphor/gaphor_plugin_helloworld)).

It's a lengthy set of rules you shared, so it may not be feasible to to validation instantly.

And how to present the validation results? As a list in a separate window?

Another idea for validation is to define those in CSS. Although I expect that will only work for only the simple rules.

**Reply by mikekidner** [reply]
Thanks for your suggestions, I agree that a command line interface so that a validation CI job can be run is the right way to go. 
CSS for some simple rules is appealing from a user point of view, but you're right that it would ultimately be quite limiting.
My initial thoughts for a validator class look like this, where rules are a collection of lambda functions which are created by parsing a json, and the validator is specialised for each element type:
![gaphor_Validator_Class_Generalizations](https://github.com/gaphor/gaphor/assets/50702106/961920ee-29e3-4493-8238-f281d4d615f3)

**Comment by mikekidner** [comment]
I've simplified the above into a checker and a list of rules. An example of rules look like this:

```
rules = [
        #Port names must start with p
        {'element_type': SysML.sysml.ProxyPort,
         'rule': "lambda e: re.match('P[a-z0-9]+', e.name)",
         'msg': "Port names must begin with the uppercase letter P"},
        #Ports must be typed by interface blocks
        {'element_type': SysML.sysml.ProxyPort,
         'rule': "lambda e: e.type.isKindOf(gaphor.SysML.sysml.InterfaceBlock)",
         'msg':"Port must be typed by an interface block"},
        #Block names must start with upper case and no spaces
        {'element_type': UML.Class,
         'rule': "lambda e: re.match('([A-Z]+[a-z0-9]+)+', e.name)",
         'msg': "Block names must begin with the uppercase letter"},        
        #Diagram names must not be the default
        {'element_type': SysML.sysml.Diagram,
         'rule': "lambda e: 'New Diagram' != e.name",
         'msg': "Diagrams must not have default name"},        
        #Activity names must not be the default
        {'element_type': UML.Activity,
         'rule': "lambda e: 'New Activity' != e.name",
         'msg': "Activities must not have default name"},        
        #Call Behaviour actions must reference an action
        {'element_type': UML.uml.CallBehaviorAction,
         'rule': "lambda e: e.behavior is not None",
         'msg': "Call behaviors must reference a behaviour"},        
        #Decision node outputs must have guards
        {'element_type': UML.DecisionNode,
         'rule': "lambda e: np.logical_and([c.guard != None for c in e._outgoing])",
         'msg': "All outputs of a Decision Node must have a guard"},        

    ]
```
The intent is that the rule set can be saved as json, making it user configurable. There are probably nicer ways than this that avoid calling eval() on a lambda string, but this seems to work well. Each rule becomes a ValidationRule, and then the apply() method evaluates the rule itself.

like this:
```
    VR = vt.Validator(validate_model, rules)
    for r in rules[-1:]: #looping over every rule, pull the relevant element type.
        for e in VR.factory.lselect(r['element_type']):
            result = VR.applyRules(e)
            print(result)
```


The ValidationRule object looks like this:

```
class Validation_Rule: 
    def __init__(self,rule, element_type, msg=""):
        self.rule = rule
        self.element_type = element_type
        self.msg = msg

    def apply(self, e):
        if e.isKindOf(self.element_type):
            print(f"Evaluating rules for {e.qualifiedName}.\nRule is: {self.rule}")
            rule = eval(self.rule)
            try:
                result = bool(rule(e))
            except:
                result = False
            if not result:
                print(self.msg)
        else:
            result = None
        return result
```

Any comments on this approach or scalability to a plugin would be welcome.
